homeassistant/components/google_assistant/trait.py

```diff
@@ -1606,41 +1607,56 @@ async def execute(self, command, data, params, challenge):
 
 @register_trait
 class VolumeTrait(_Trait):
-    """Trait to control brightness of a device.
+    """Trait to control volume of a device.
 
     https://developers.google.com/actions/smarthome/traits/volume
     """
 
     name = TRAIT_VOLUME
-    commands = [COMMAND_SET_VOLUME, COMMAND_VOLUME_RELATIVE]
+    commands = [COMMAND_SET_VOLUME, COMMAND_VOLUME_RELATIVE, COMMAND_MUTE]
 
     @staticmethod
     def supported(domain, features, device_class):
-        """Test if state is supported."""
+        """Test if trait is supported."""
         if domain == media_player.DOMAIN:
-            return features & media_player.SUPPORT_VOLUME_SET
+            return features & (
+                media_player.SUPPORT_VOLUME_SET | media_player.SUPPORT_VOLUME_STEP
+            )
 
         return False
 
     def sync_attributes(self):
-        """Return brightness attributes for a sync request."""
-        return {}
+        """Return volume attributes for a sync request."""
+        features = self.state.attributes.get(ATTR_SUPPORTED_FEATURES, 0)
+        return {
+            "volumeMaxLevel": 100,
+            "volumeCanMuteAndUnmute": bool(features & media_player.SUPPORT_VOLUME_MUTE),
+            "commandOnlyVolume": not (features & media_player.SUPPORT_VOLUME_SET),
```

Comment:
This is wrong. If we dont support volume set, we don't support setting of volume. An entity can still support reporting volume level, but only support changing by step method.

This should be true for a assume state entity.

Reply:
Really, what needs to happen here is we should be checking whether the `media_player` supports querying volume. The `commandOnlyVolume` attribute, as far as I can tell, is only used to determine whether to send `QUERY` requests for the trait. Despite what the documentation says, it still sends relative-volume requests even for non-`commandOnlyVolume` devices. The consequences of getting it wrong would be:

- `commandOnlyVolume` set when it shouldn't be: Google Assistant won't be able to respond to queries for the device's current volume.
- `commandOnlyVolume` not set when it should be: We won't set `currentVolume` in the `QUERY` response, and explicit That's marked as required, so it probably means the trait update gets tossed as invalid, but if we don't have volume to query, there's not really a state to update anyway.

Ideally, there'd be a SUPPORT_VOLUME_READ bit to determine whether we could safely read volume, but without it, setting `commandOnlyVolume` based on `ATTR_ASSUMED_STATE` is an overestimate, but OK. The consequences of setting `commandOnlyVolume` to `false` for something it shouldn't be isn't awful (after all, that's been the case up until now), but setting it to `true` for something that could support it means we lose functionality.

I'll go ahead and make that change.

Reply:
Being able to query is what assumed state attribute is all about. I agree that there might be devices that don't set that value and still can't query volume. But it's now our best bet.

Reply:
The most recent version of this commit has that change in place. Please take a look.